### PR TITLE
[SPARK-15812][SQ][Streaming] Added support for sorting after streaming aggregation with complete mode

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

When the output mode is complete, then the output of a streaming aggregation essentially will contain the complete aggregates every time. So this is not different from a batch dataset within an incremental execution. Other non-streaming operations should be supported on this dataset. In this PR, I am just adding support for sorting, as it is a common useful functionality. Support for other operations will come later.
## How was this patch tested?

Additional unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala [hunk 1]
```diff
@@ -123,27 +159,6 @@ object UnsupportedOperationChecker {
         case _ =>
       }
     }
-
-    // Checks related to aggregations
-    val aggregates = plan.collect { case a @ Aggregate(_, _, _) if a.isStreaming => a }
```
Comment (tdas) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: This is moved above to make sure that outputmode related failure occur before other failures, as those failures are more fundamental.

**Comment by tdas** [comment]
@marmbrus

**Comment by rxin** [comment]
Any reason why there isn't a jira ticket?

**Comment by tdas** [comment]
Added. My bad.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala [hunk 2]
```diff
@@ -43,6 +43,41 @@ object UnsupportedOperationChecker {
...
+      case _ =>
+    }
+
+    /**
+     * Whether the subplan will contain complete data or incremental data in every incremental
+     * execution. Some operations may be allowed only when the child logical plan gives complete
+     * data.
+     */
+    def containsCompleteData(subplan: LogicalPlan): Boolean = {
+      val aggs = plan.collect { case a@Aggregate(_, _, _) if a.isStreaming => a }
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: There is a lot of implicit coupling here.  I think the correctness of this function depends on the fact that:
 - we only allow aggregation
 - currently the aggregate operator is responsible for outputting complete data
 - aggregates always output complete data if the mode is complete

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala [hunk 3]
```diff
@@ -104,6 +104,31 @@ class StreamingAggregationSuite extends StreamTest with BeforeAndAfterAll {
     }
   }
 
+  test("operation after aggregate in complete mode") {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala]: what operation

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 4]
```diff
@@ -189,9 +189,20 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
     _.intersect(_),
     streamStreamSupported = false)
 
-  // Unary operations
+  // Sort: supported only on batch subplans and on aggregation + complete output mode
   testUnaryOperatorInStreamingPlan("sort", Sort(Nil, true, _))
   testUnaryOperatorInStreamingPlan("sort partitions", SortPartitions(Nil, _), expectedMsg = "sort")
+  assertSupportedInStreamingPlan(
+    "sort - sort over aggregated data in Complete output mode",
+    Sort(Nil, true, Aggregate(Nil, aggExprs("c"), streamRelation)),
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala]: I think these would be easier to read written with the DSL.

**Comment by marmbrus** [comment]
This is okay for 2.0, but we'll need to rethink the way we are doing query planning to handle incremental input.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala (hunk 2 above)
Comment (tdas) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: I completely agree. This whole file needs to be restructured in future. This is really a 2.0 interim fix.

**Comment by yhuai** [comment]
test this please

**Comment by tdas** [comment]
@marmbrus could you take a look once again.

**Pull request closed without merging**